# 개요
https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce

Brazilian E-Commerce Public Dataset by Olist : Olist 브라질 이커머스 데이터셋 EDA

<img src='https://i.imgur.com/HRhd2Y0.png' width=700>

In [18]:
import pandas as pd

#관계

##1. customer, sellers & geolocation 관계
- sellers : seller_zip_code_prefit
- customer : customer_zip_code_prefix
- geolocation : geolocation_zip_code_prefix
-> XXX_zip_code_prefix로 관계를 이루고 있음.



In [19]:
df_seller = pd.read_csv('olist_sellers_dataset.csv')
df_cust = pd.read_csv('olist_customers_dataset.csv')
df_geo = pd.read_csv('olist_geolocation_dataset.csv')

In [20]:
df_geo.head(3)

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP


In [21]:
df_cust.head(3)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP


In [22]:
df_seller.head(3)

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ


In [12]:
df_geo[''].drop_duplicates()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP
...,...,...,...,...,...
1000155,99965,-28.180655,-52.034367,agua santa,RS
1000156,99950,-28.072188,-52.011272,tapejara,RS
1000157,99950,-28.068864,-52.012964,tapejara,RS
1000158,99950,-28.068639,-52.010705,tapejara,RS


In [10]:
pd.merge(df_cust, df_geo, left_on='customer_zip_code_prefix', right_on='geolocation_zip_code_prefix', how='left')

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP,14409.0,-20.509897,-47.397866,franca,SP
1,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP,14409.0,-20.497396,-47.399241,franca,SP
2,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP,14409.0,-20.510459,-47.399553,franca,SP
3,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP,14409.0,-20.480940,-47.394161,franca,SP
4,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP,14409.0,-20.515413,-47.398194,franca,SP
...,...,...,...,...,...,...,...,...,...,...
15083728,274fa6071e5e17fe303b9748641082c8,84732c5050c01db9b23e19ba39899398,6703,cotia,SP,6703.0,-23.599369,-46.905603,cotia,SP
15083729,274fa6071e5e17fe303b9748641082c8,84732c5050c01db9b23e19ba39899398,6703,cotia,SP,6703.0,-23.593577,-46.910112,cotia,SP
15083730,274fa6071e5e17fe303b9748641082c8,84732c5050c01db9b23e19ba39899398,6703,cotia,SP,6703.0,-23.584425,-46.892014,cotia,SP
15083731,274fa6071e5e17fe303b9748641082c8,84732c5050c01db9b23e19ba39899398,6703,cotia,SP,6703.0,-23.595022,-46.918546,cotia,SP


##2. orders + order_payments & order_reviews & order_items

In [44]:
df_ord = pd.read_csv('olist_orders_dataset.csv')
df_ord_pay = pd.read_csv('olist_order_payments_dataset.csv')
df_ord_rev = pd.read_csv('olist_order_reviews_dataset.csv')
df_ord_itm = pd.read_csv('olist_order_items_dataset.csv')

In [45]:
print(df_ord.shape, df_ord_pay.shape, df_ord_rev.shape, df_ord_itm.shape)

(99441, 8) (103886, 5) (99224, 7) (112650, 7)


In [14]:
df_ord.head(3)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00


In [16]:
df_ord_pay.head(3)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71


In [25]:
df_ord_rev.head(3)

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24


In [46]:
df_ord_itm.head(3)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.0,17.87


- ord & ord_pay : 하나의 주문을 여러 결제수단으로 나누어 결제하는 경우 order_id의 중복이 생김.

In [23]:
pd.merge(left = df_ord, right = df_ord_pay, on='order_id', how='left')

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,payment_sequential,payment_type,payment_installments,payment_value
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,1.0,credit_card,1.0,18.12
1,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,3.0,voucher,1.0,2.00
2,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,2.0,voucher,1.0,18.59
3,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00,1.0,boleto,1.0,141.46
4,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00,1.0,credit_card,3.0,179.12
...,...,...,...,...,...,...,...,...,...,...,...,...
103882,9c5dedf39a927c1b2549525ed64a053c,39bd1228ee8140590ac3aca26f2dfe00,delivered,2017-03-09 09:54:05,2017-03-09 09:54:05,2017-03-10 11:18:03,2017-03-17 15:08:01,2017-03-28 00:00:00,1.0,credit_card,3.0,85.08
103883,63943bddc261676b46f01ca7ac2f7bd8,1fca14ff2861355f6e5f14306ff977a7,delivered,2018-02-06 12:58:58,2018-02-06 13:10:37,2018-02-07 23:22:42,2018-02-28 17:37:56,2018-03-02 00:00:00,1.0,credit_card,3.0,195.00
103884,83c1379a015df1e13d02aae0204711ab,1aa71eb042121263aafbe80c1b562c9c,delivered,2017-08-27 14:46:43,2017-08-27 15:04:16,2017-08-28 20:52:26,2017-09-21 11:24:17,2017-09-27 00:00:00,1.0,credit_card,5.0,271.01
103885,11c177c8e97725db2631073c19f07b62,b331b74b18dc79bcdf6532d51e1637c1,delivered,2018-01-08 21:28:27,2018-01-08 21:36:21,2018-01-12 15:35:03,2018-01-25 23:32:54,2018-02-15 00:00:00,1.0,credit_card,4.0,441.16


- ord & ord_rev : order_id 별 review 2개 이상인것 꽤 있음. 하지만 65%는 review 값이 NaN. 아마 오류로 NaN 들어간 다음에 정상 review가 들어간 것 같음.

In [42]:
temp = pd.merge(left = df_ord, right = df_ord_rev, on='order_id', how='left')
temp_dup = temp[temp['order_id'].duplicated(keep=False)].sort_values(by='order_id')
sum(temp_dup['review_comment_message'].isnull()) / len(temp_dup)

0.6493624772313297

In [43]:
temp_dup.head(10)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,review_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
91717,0035246a40f520710769010f752e7507,4bcf2ad7ccc061c13ff360aae5b7abcd,delivered,2017-08-17 00:57:23,2017-08-17 01:25:39,2017-08-22 15:33:53,2017-08-28 19:14:03,2017-09-06 00:00:00,89a02c45c340aeeb1354a24e7d4b2c1e,5.0,NaN,NaN,2017-08-29 00:00:00,2017-08-30 01:59:12
91716,0035246a40f520710769010f752e7507,4bcf2ad7ccc061c13ff360aae5b7abcd,delivered,2017-08-17 00:57:23,2017-08-17 01:25:39,2017-08-22 15:33:53,2017-08-28 19:14:03,2017-09-06 00:00:00,2a74b0559eb58fc1ff842ecc999594cb,5.0,NaN,Estou acostumada a comprar produtos pelo barat...,2017-08-25 00:00:00,2017-08-29 21:45:57
84704,013056cfe49763c6f66bda03396c5ee3,d3d38825275e47847abef7e1936b3bcb,delivered,2018-02-14 12:48:24,2018-02-14 12:55:56,2018-02-15 20:32:36,2018-03-03 12:09:00,2018-03-07 00:00:00,73413b847f63e02bc752b364f6d05ee9,4.0,NaN,NaN,2018-03-04 00:00:00,2018-03-05 17:02:00
84703,013056cfe49763c6f66bda03396c5ee3,d3d38825275e47847abef7e1936b3bcb,delivered,2018-02-14 12:48:24,2018-02-14 12:55:56,2018-02-15 20:32:36,2018-03-03 12:09:00,2018-03-07 00:00:00,ab30810c29da5da8045216f0f62652a2,5.0,NaN,NaN,2018-02-22 00:00:00,2018-02-23 12:12:30
94711,0176a6846bcb3b0d3aa3116a9a768597,aaad67429ea7d339bdf5f71ae3ef74f6,delivered,2017-12-25 10:40:52,2017-12-27 04:32:07,2017-12-28 19:57:15,2017-12-29 21:22:35,2018-01-15 00:00:00,830636803620cdf8b6ffaf1b2f6e92b2,5.0,NaN,NaN,2017-12-30 00:00:00,2018-01-02 10:54:06
94712,0176a6846bcb3b0d3aa3116a9a768597,aaad67429ea7d339bdf5f71ae3ef74f6,delivered,2017-12-25 10:40:52,2017-12-27 04:32:07,2017-12-28 19:57:15,2017-12-29 21:22:35,2018-01-15 00:00:00,d8e8c42271c8fb67b9dad95d98c8ff80,5.0,NaN,NaN,2017-12-30 00:00:00,2018-01-02 10:54:47
81030,02355020fd0a40a0d56df9f6ff060413,b28f05c74142c8e7c6e07e5336fa63ae,delivered,2018-02-28 16:07:09,2018-03-02 02:15:42,2018-03-02 19:16:33,2018-04-10 14:48:49,2018-03-26 00:00:00,0c8e7347f1cdd2aede37371543e3d163,3.0,NaN,UM DOS PRODUTOS (ENTREGA02) COMPRADOS NESTE PE...,2018-03-21 00:00:00,2018-03-22 01:32:08
81029,02355020fd0a40a0d56df9f6ff060413,b28f05c74142c8e7c6e07e5336fa63ae,delivered,2018-02-28 16:07:09,2018-03-02 02:15:42,2018-03-02 19:16:33,2018-04-10 14:48:49,2018-03-26 00:00:00,017f0e1ea6386de662cbeba299c59ad1,1.0,NaN,ja reclamei varias vezes e ate hoje não sei on...,2018-03-29 00:00:00,2018-03-30 03:16:19
73790,029863af4b968de1e5d6a82782e662f5,74179773f1f31f3af11c3b3b065115f1,delivered,2017-07-06 09:03:55,2017-07-06 09:15:11,2017-07-10 19:43:43,2017-07-13 20:42:36,2017-07-31 00:00:00,61fe4e7d1ae801bbe169eb67b86c6eda,4.0,NaN,NaN,2017-07-19 00:00:00,2017-07-20 12:06:11
73789,029863af4b968de1e5d6a82782e662f5,74179773f1f31f3af11c3b3b065115f1,delivered,2017-07-06 09:03:55,2017-07-06 09:15:11,2017-07-10 19:43:43,2017-07-13 20:42:36,2017-07-31 00:00:00,04d945e95c788a3aa1ffbee42105637b,5.0,NaN,NaN,2017-07-14 00:00:00,2017-07-17 13:58:06


##3. items & products

In [67]:
df_prod = pd.read_csv('olist_products_dataset.csv')
df_prod.shape

(32951, 9)

In [61]:
df_prod.head(3)

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0


In [62]:
df_ord_itm.head(3)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.0,17.87


In [70]:
temp = pd.merge(df_ord_itm, df_prod, on='product_id', how='left')
temp[temp['product_id'].duplicated(keep=False)].sort_values('product_id')

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
48878,6f0dfb5b5398b271cc6bbd9ee263530e,1,000b8f95fcb9e0096488278317764d19,40ec8ab6cdafbcc4f544da38c67da39a,2018-08-07 22:15:19,58.90,19.60,utilidades_domesticas,25.0,364.0,3.0,550.0,19.0,24.0,12.0
25833,3aba44d8e554ab4bb8c09f6f78032ca8,1,000b8f95fcb9e0096488278317764d19,40ec8ab6cdafbcc4f544da38c67da39a,2018-08-16 13:35:21,58.90,19.60,utilidades_domesticas,25.0,364.0,3.0,550.0,19.0,24.0,12.0
93674,d44be4540d1a35533a9f5c89377b77f1,1,00126f27c813603687e6ce486d909d01,cd68562d3f44870c08922d380acae552,2017-09-21 20:50:13,249.00,16.49,cool_stuff,42.0,2461.0,1.0,700.0,25.0,5.0,15.0
88768,c9a34c30282dc36d017bc912f8d8dc91,1,00126f27c813603687e6ce486d909d01,cd68562d3f44870c08922d380acae552,2017-09-21 20:55:19,249.00,13.24,cool_stuff,42.0,2461.0,1.0,700.0,25.0,5.0,15.0
87904,c78ef951d29d8e1af5f1f807a64af309,1,001795ec6f1b187d37335e1c4704762e,8b321bb669392f5163d04c59e235e066,2018-01-03 00:34:19,38.90,9.34,consoles_games,53.0,274.0,1.0,600.0,30.0,20.0,20.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
97508,dd38c566a34ff9deb93ff05d7bd6d0aa,1,fffdb2d0ec8d6a61f0a0a0db3f25b441,9f505651f4a6abe901a56cdc21508025,2018-05-07 17:53:01,34.99,7.39,informatica_acessorios,47.0,501.0,5.0,150.0,16.0,4.0,13.0
94694,d6c1ed12e8eeaa15e9d5c83c79598f85,1,fffdb2d0ec8d6a61f0a0a0db3f25b441,9f505651f4a6abe901a56cdc21508025,2018-08-14 17:25:18,34.99,7.51,informatica_acessorios,47.0,501.0,5.0,150.0,16.0,4.0,13.0
83941,beaf95465b542dea76df78b2fb86a97a,1,fffdb2d0ec8d6a61f0a0a0db3f25b441,9f505651f4a6abe901a56cdc21508025,2018-04-25 10:30:35,34.99,18.23,informatica_acessorios,47.0,501.0,5.0,150.0,16.0,4.0,13.0
73757,a8159270bc9cf7d54450e079ba6b7232,1,fffdb2d0ec8d6a61f0a0a0db3f25b441,9f505651f4a6abe901a56cdc21508025,2018-04-09 05:35:27,29.99,18.23,informatica_acessorios,47.0,501.0,5.0,150.0,16.0,4.0,13.0


#단일

- geolocation : zip_code와 그 위도와 경도, 도시와 주. 같은 zip_code_prefix라도 위도, 경도가 조금씩 다른 데이터가 많음. zip code는 같아도 그 안에 주소가 많을 수 있기 때문.
index|geolocation_zip_code_prefix|geolocation_lat|geolocation_lng|geolocation_city|geolocation_state|
|---|---|---|---|---|---|
1|1046|-23.546081|-46.644820|sao paulo|SP|
2|1046|-23.546129|-46.642951|sao paulo|SP|

In [71]:
df_geo

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP
...,...,...,...,...,...
1000158,99950,-28.068639,-52.010705,tapejara,RS
1000159,99900,-27.877125,-52.224882,getulio vargas,RS
1000160,99950,-28.071855,-52.014716,tapejara,RS
1000161,99980,-28.388932,-51.846871,david canabarro,RS


In [78]:
df_cust['customer_unique_id'].duplicated()

,customer_unique_id
0,False
1,False
2,False
3,False
4,False
...,...
99436,False
99437,False
99438,False
99439,False
